examples/vision/siamesenetwork.py

```diff
@@ -0,0 +1,393 @@
+
+
+![1_0E9104t29iMBmtvq7G1G6Q.png](attachment:1_0E9104t29iMBmtvq7G1G6Q.png)
+"""
+
+"""
+First we get the paths of the datasets in siamese networks we usually have two folders
+each folder has images and every image has a corresponding similar picture in the other
+folder.
+"""
+
+
+def norm(embeddings):
+    """
+    Computes the euclidean norm of the three embeddings generated by the
+    Siamese network.
+
+    The first supplied embedding corresponds to the anchor image. The second
+    embedding corresponds to the positive image, and the third embedding
+    corresponds to the negative image.
+    """
+
+    anchor_positive_dist = tf.norm(embeddings[0] - embeddings[1], axis=1, keepdims=True)
+    anchor_negative_dist = tf.norm(embeddings[0] - embeddings[2], axis=1, keepdims=True)
+
+    return (anchor_positive_dist, anchor_negative_dist)
+
+
+"""
+## Model subclassing
+
+Here we customize our training process and our model.
+
+We override the train_step() method and apply our own loss and our own training process
+
+We also use Triplet loss function as we specified above.
+
+Loss function explaination:
+
+we calculate the distance between the anchor embedding and the positive embedding the
+axis = -1 because we want the distance over the features of every example. We also add
+alpha which act as extra margin.
+"""
+
+
+class SiameseModel(Model):
+    def __init__(self, siamese_network, alpha=0.5):
+        super(SiameseModel, self).__init__()
+        self.siamese_network = siamese_network
+        self.alpha = alpha
+
+    def call(self, inputs):
+        pass
+
+    def train_step(self, data):
+        # here we create a tape to record our operations so we can get the gradients
+        with tf.GradientTape() as tape:
+            anchor, positive, negative = data
+            distances = self.siamese_network((anchor, positive, negative))
+
+            # getting the loss by subtracting the distances
+            loss = distances[0] - distances[1]
+
+            # getting the max because we don't want negative loss
+            loss = tf.reduce_mean(tf.maximum(loss + self.alpha, 0.0))
+
+        # getting the gradients [loss with respect to trainable weights]
+        grads = tape.gradient(loss, self.siamese_network.trainable_weights)
+
+        # applying the gradients on the model using the specified optimizer
+        self.optimizer.apply_gradients(
+            zip(grads, self.siamese_network.trainable_weights)
+        )
+
+        return {"loss": loss}
+
+
+"""
+## Cosine Similarity Layer
+
+This layer just computes how similar to feature vectors are by computing it using the
+Cosine Similarity
+We override the call method and implement our own call method.
+
+Check out https://www.tensorflow.org/api_docs/python/tf/keras/losses/CosineSimilarity
+
+We return the negative of the loss because we just need to know how similar they are we
+do NOT need to know the loss
+"""
+
+
+class CosineDistance(layers.Layer):
```

Comment:
In general, define each new object in the section that will need it. This object only gets used at line 336. Same for other classes/functions like `norm`, etc.